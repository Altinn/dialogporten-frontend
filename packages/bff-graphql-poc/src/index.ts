#!/usr/bin/env -S node --no-warnings=ExperimentalWarning --loader ts-node/esm
import debug from 'debug';
import Fastify from 'fastify';
import fastifyGraphiql from 'fastify-graphiql';
import debugScreen from './debug-screen.tsx';
import { host, mode, port } from './env.ts';
import fastifyGraphql from './graphql.ts';

// Setting up logger function
const log = debug('app');

// Setting up webserver
const app = Fastify();

app.register(fastifyGraphql);

if (mode === 'development') {
  log('`development` mode, activating Graphiql and debug screen');
  app.register(fastifyGraphiql);
  app.register(debugScreen);
}

app.listen({ port, host }, (err, address) => {
  if (err) {
    log(err);
    process.exit(1);
  }
  console.log(`App started on ${address}`);
});
