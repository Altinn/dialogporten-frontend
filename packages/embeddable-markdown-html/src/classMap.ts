export type ClassMap = Partial<Record<string, string>>;

export const defaultClassMap: ClassMap = {
  p: '__fce-paragraph',
  li: '__fce-list-item',
  h1: '__fce-header-1',
  h2: '__fce-header-2',
  h3: '__fce-header-3',
  h4: '__fce-header-4',
  h5: '__fce-header-5',
  h6: '__fce-header-6',
  strong: '__fce-strong',
};
