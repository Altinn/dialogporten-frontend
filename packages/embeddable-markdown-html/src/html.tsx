import { type ReactElement, useEffect, useState } from 'react';
import * as prod from 'react/jsx-runtime';
import rehypeParse, { type Options as RehypeParseOptions } from 'rehype-parse';
import rehypeReact from 'rehype-react';
import { unified } from 'unified';

// @ts-expect-error: the react types are missing.
const production = { Fragment: prod.Fragment, jsx: prod.jsx, jsxs: prod.jsxs };

export const Html: React.FC<{
  children: string;
  onError?: (err: Error) => void;
}> = ({ children, onError = () => {} }) => {
  const [reactContent, setReactContent] = useState<ReactElement | null>(null);

  // TODO: Could this be a useMemo call?
  useEffect(() => {
    unified()
      .use(rehypeParse, {} as RehypeParseOptions)
      .use(rehypeReact, production)
      .process(children)
      .then((vfile) => setReactContent(vfile.result as ReactElement))
      .catch((e) => onError(e));
  }, [children, onError]);

  return reactContent;
};
