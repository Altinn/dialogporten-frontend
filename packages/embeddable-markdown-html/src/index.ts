export * from './markdown.tsx';
export * from './html.tsx';
