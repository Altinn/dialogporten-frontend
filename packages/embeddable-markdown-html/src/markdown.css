:root {
  --markdown-base-size: 1.125rem;
  --markdown-base-line-height: 1.5;
}
.__markdown-paragraph {
  font-size: var(--markdown-base-size);
  line-height: var(--markdown-base-line-height);
}

.__markdown-list-item {
  font-size: var(--markdown-base-size);
  line-height: var(--markdown-base-line-height);
}

.__markdown-header-3,
.__markdown-header-4,
.__markdown-header-5,
.__markdown-header-6 {
  font-size: var(--markdown-base-size);
  line-height: var(--markdown-base-line-height);
}

.__markdown-header-1,
.__markdown-header-2,
.__markdown-header-3,
.__markdown-header-4,
.__markdown-header-5,
.__markdown-header-6 {
  margin: 0.75rem 0;
}

.__markdown-header-1 {
  font-size: 1.5rem;
  line-height: 1.25;
}

.__markdown-header-1 {
  font-weight: 500;
}

.__markdown-header-2 {
  font-weight: 500;
  font-size: 1.25rem;
  line-height: 1.25;
}

.__markdown-header-3 {
  font-weight: 400;
  font-size: 1.25rem;
  line-height: 1.25;
}

.__markdown-header-4 {
  font-weight: 600;
}

.__markdown-header-5 {
  font-weight: 500;
}

.__markdown-header-6 {
  font-weight: 500;
}

.__markdown-strong {
  font-weight: 500;
}
