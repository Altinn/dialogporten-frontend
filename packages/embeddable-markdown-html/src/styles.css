:root {
  --fce-base-size: 1.125rem;
  --fce-base-line-height: 1.5;
}

.__fce-paragraph {
  font-size: var(--fce-base-size);
  line-height: var(--fce-base-line-height);
}

.__fce-list-item {
  font-size: var(--fce-base-size);
  line-height: var(--fce-base-line-height);
}

.__fce-header-3,
.__fce-header-4,
.__fce-header-5,
.__fce-header-6 {
  font-size: var(--fce-base-size);
  line-height: var(--fce-base-line-height);
}

.__fce-header-1,
.__fce-header-2,
.__fce-header-3,
.__fce-header-4,
.__fce-header-5,
.__fce-header-6 {
  margin: 0.75rem 0;
}

.__fce-header-1 {
  font-size: 1.5rem;
  line-height: 1.25;
  font-weight: 500;
}

.__fce-header-2 {
  font-weight: 500;
  font-size: 1.25rem;
  line-height: 1.25;
}

.__fce-header-3 {
  font-weight: 400;
  font-size: 1.25rem;
  line-height: 1.25;
}

.__fce-header-4 {
  font-weight: 600;
}

.__fce-header-5 {
  font-weight: 500;
}

.__fce-header-6 {
  font-weight: 500;
}

.__fce-strong {
  font-weight: 500;
}
