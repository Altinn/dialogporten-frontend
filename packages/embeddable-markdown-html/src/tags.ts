export const allowedTags = [
  'p',
  'em',
  'strong',
  'ul',
  'ol',
  'li',
  'a',
  'code',
  'pre',
  'blockquote',
  'hr',
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'br',
  'span',
];
