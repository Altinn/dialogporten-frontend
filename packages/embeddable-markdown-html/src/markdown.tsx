import type { Options as RemarkRehypeOptions } from 'mdast-util-to-hast';
import { type ReactElement, useEffect, useState } from 'react';
import * as prod from 'react/jsx-runtime';
import addClasses from 'rehype-class-names';
import rehypeReact from 'rehype-react';
import rehypeSanitize from 'rehype-sanitize';
import remarkParse, { type Options as RemarkParseOptions } from 'remark-parse';
import remarkToRehype from 'remark-rehype';
import { unified } from 'unified';

import './markdown.css';

type ClassMap = Partial<Record<string, string>>;
// @ts-expect-error: the react types are missing.
const production = { Fragment: prod.Fragment, jsx: prod.jsx, jsxs: prod.jsxs };

const defaultClassMap: ClassMap = {
  p: '__markdown-paragraph',
  li: '__markdown-list-item',
  h1: '__markdown-header-1',
  h2: '__markdown-header-2',
  h3: '__markdown-header-3',
  h4: '__markdown-header-4',
  h5: '__markdown-header-5',
  h6: '__markdown-header-6',
  strong: '__markdown-strong',
};

/**
 * Renders markdown as React elements.
 *
 * @param children - The markdown string to render.
 * @param onError - A callback for handling errors.
 * @param classMap - A map of HTML element names to CSS classes.
 * @returns The rendered React elements.
 */
export const Markdown: ({
  children,
  onError,
  classMap,
}: { children: string; onError?: (error: unknown) => void; classMap?: ClassMap }) => ReactElement | null = ({
  children,
  onError = () => {},
  classMap = defaultClassMap,
}) => {
  const [reactContent, setReactContent] = useState<ReactElement | null>(null);

  // biome-ignore lint/correctness/useExhaustiveDependencies: Full control of what triggers this code is needed
  useEffect(() => {
    unified()
      .use(remarkParse, {} as RemarkParseOptions)
      .use(remarkToRehype, {} as RemarkRehypeOptions)
      .use(rehypeSanitize)
      .use(addClasses, classMap)
      .use(rehypeReact, production)
      .process(children)
      .then((vfile) => setReactContent(vfile.result as ReactElement))
      .catch(onError);
  }, [children]);

  return reactContent;
};
