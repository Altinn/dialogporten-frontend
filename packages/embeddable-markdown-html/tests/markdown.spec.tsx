import { render, waitFor } from '@testing-library/react';
import { describe, expect, test } from 'vitest';
import { Markdown } from '../src';

describe('Markdown: common mark + table support', () => {
  /* Tests for supported tags */
  test('should render paragraph', async () => {
    const markdownString = 'This is a paragraph.';
    const { container } = render(<Markdown onError={() => {}}>{markdownString}</Markdown>);
    await waitFor(() => {
      expect(container).toHaveTextContent('This is a paragraph.');
    });
    expect(container.innerHTML).toEqual('<p class="__fce-paragraph">This is a paragraph.</p>');
  });
  test('should render italic text', async () => {
    const markdownString = '*italic*';
    const { container } = render(<Markdown onError={() => {}}>{markdownString}</Markdown>);
    await waitFor(() => {
      expect(container.innerHTML).toContain('<em>italic</em>');
    });
  });
  test('should render bold text', async () => {
    const markdownString = '**bold**';
    const { container } = render(<Markdown onError={() => {}}>{markdownString}</Markdown>);
    await waitFor(() => {
      expect(container.innerHTML).toContain('<strong class="__fce-strong">bold</strong>');
    });
  });
  test('should render unordered list', async () => {
    const markdownString = '- Item 1\n- Item 2';
    const { container } = render(<Markdown onError={() => {}}>{markdownString}</Markdown>);
    await waitFor(() => {
      expect(container.querySelectorAll('ul > li').length).toBe(2);
    });
  });
  test('should render ordered list', async () => {
    const markdownString = '1. First\n2. Second';
    const { container } = render(<Markdown onError={() => {}}>{markdownString}</Markdown>);
    await waitFor(() => {
      expect(container.querySelectorAll('ol > li').length).toBe(2);
    });
  });
  test('should render link', async () => {
    const markdownString = '[Altinn](https://altinn.no)';
    const { container } = render(<Markdown onError={() => {}}>{markdownString}</Markdown>);
    await waitFor(() => {
      const link = container.querySelector('a');
      expect(link).toHaveAttribute('href', 'https://altinn.no');
      expect(link).toHaveTextContent('Altinn');
    });
  });
  test('should render inline code', async () => {
    const markdownString = 'Here is `inline code`.';
    const { container } = render(<Markdown onError={() => {}}>{markdownString}</Markdown>);
    await waitFor(() => {
      expect(container.innerHTML).toContain('<code>inline code</code>');
    });
  });
  test('should render code block', async () => {
    const markdownString = '```\nconst x = 1;\n```';
    const { container } = render(<Markdown onError={() => {}}>{markdownString}</Markdown>);
    await waitFor(() => {
      expect(container.querySelector('pre > code')).toHaveTextContent('const x = 1;');
    });
  });
  test('should render blockquote', async () => {
    const markdownString = '> This is a quote.';
    const { container } = render(<Markdown onError={() => {}}>{markdownString}</Markdown>);
    await waitFor(() => {
      expect(container.querySelector('blockquote')).toHaveTextContent('This is a quote.');
    });
  });
  test('should render horizontal rule', async () => {
    const markdownString = '---';
    const { container } = render(<Markdown onError={() => {}}>{markdownString}</Markdown>);
    await waitFor(() => {
      expect(container.querySelector('hr')).toBeInTheDocument();
    });
  });
  test('should render headings', async () => {
    const markdownString = '# H1\n## H2\n### H3\n#### H4\n##### H5\n###### H6';
    const { container } = render(<Markdown onError={() => {}}>{markdownString}</Markdown>);
    await waitFor(() => {
      for (let i = 1; i <= 6; i++) {
        expect(container.querySelector(`h${i}`)).toBeInTheDocument();
      }
    });
  });
  test('should render line break', async () => {
    const markdownString = 'Line one.  \nLine two.';
    const { container } = render(<Markdown onError={() => {}}>{markdownString}</Markdown>);
    await waitFor(() => {
      expect(container.innerHTML).toContain('<br>');
    });
  });
  test('should render paragraph', async () => {
    const markdownString = 'This is a paragraph.';
    const { container } = render(<Markdown onError={() => {}}>{markdownString}</Markdown>);
    await waitFor(() => {
      expect(container).toHaveTextContent('This is a paragraph.');
    });
    expect(container.innerHTML).toEqual('<p class="__fce-paragraph">This is a paragraph.</p>');
  });

  /* Tests for unsupported tags */
  test('should not render image', async () => {
    const markdownString = '![image example](./example.png) *something else*';
    const { container } = render(<Markdown onError={() => {}}>{markdownString}</Markdown>);
    await waitFor(() => {
      expect(container.querySelector('em')).toBeInTheDocument();
      expect(container.querySelector('img')).not.toBeInTheDocument();
    });
  });

  test('should sanitize <script> tag but render supported content', async () => {
    const markdownString = '# Heading\n<script>alert("xss")</script>';
    const { container } = render(<Markdown onError={() => {}}>{markdownString}</Markdown>);
    await waitFor(() => {
      expect(container.querySelector('h1')).toHaveTextContent('Heading');
      expect(container.innerHTML).not.toContain('<script>');
      expect(container.innerHTML).not.toContain('alert("xss")');
    });
  });

  test('should sanitize <iframe> tag but render list', async () => {
    const markdownString = '- Item\n<iframe src="https://x.com"></iframe>';
    const { container } = render(<Markdown onError={() => {}}>{markdownString}</Markdown>);
    await waitFor(() => {
      expect(container.querySelector('li')).toHaveTextContent('Item');
      expect(container.innerHTML).not.toContain('<iframe>');
      expect(container.innerHTML).not.toContain('x.com');
    });
  });

  test('should sanitize <video> tag but render bold text', async () => {
    const markdownString = '**Bold**\n<video src="foo.mp4"></video>';
    const { container } = render(<Markdown onError={() => {}}>{markdownString}</Markdown>);
    await waitFor(() => {
      expect(container.querySelector('strong')).toHaveTextContent('Bold');
      expect(container.innerHTML).not.toContain('<video>');
      expect(container.innerHTML).not.toContain('foo.mp4');
    });
  });

  test('should sanitize raw <div> HTML but render emphasized text', async () => {
    const markdownString = '*emphasis* <div>raw html</div>';
    const { container } = render(<Markdown onError={() => {}}>{markdownString}</Markdown>);
    await waitFor(() => {
      expect(container.querySelector('em')).toHaveTextContent('emphasis');
      expect(container.innerHTML).not.toContain('<div>');
    });
  });

  test('should render table', async () => {
    const markdownString = `| Syntax      | Description |
| ----------- | ----------- |
| Header      | Title       |
| Paragraph   | Text        |`;
    const { container } = render(<Markdown onError={() => {}}>{markdownString}</Markdown>);
    await waitFor(() => {
      expect(container.querySelector('table')).toHaveTextContent('Syntax');
      expect(container.querySelector('table')).toHaveTextContent('Description');
      expect(container.querySelectorAll('td')[0]).toHaveTextContent('Header');
      expect(container.querySelectorAll('td')[1]).toHaveTextContent('Title');
      expect(container.querySelectorAll('td')[2]).toHaveTextContent('Paragraph');
      expect(container.querySelectorAll('td')[3]).toHaveTextContent('Text');
    });
  });
});
