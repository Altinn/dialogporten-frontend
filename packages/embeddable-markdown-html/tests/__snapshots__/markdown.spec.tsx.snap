// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Markdown > should render content 1`] = `
<h1
  class="__markdown-header-1"
>
  header
</h1>
`;
